wnfs/src/private/share.rs

```diff
@@ -0,0 +1,508 @@
+        Self {
+            payload,
+            count,
+            sharer: None,
+            recipients: Vec::new(),
+            phantom: PhantomData,
+        }
+    }
+
+    /// Sets the sharer field.
+    pub fn by(&mut self, sharer: Sharer<'a, S>) -> &mut Self {
+        self.sharer = Some(sharer);
+        self
+    }
+
+    /// Takes a vector of recipients as an argument, and adds it to the existing recipients of the Share struct.
+    pub fn with(&mut self, recipients: Vec<Recipient<'a, S>>) -> &mut Self {
+        self.recipients.extend(recipients);
+        self
+    }
+
+    /// Takes a recipient as an argument and adds it to the existing recipients of the Share struct.
+    pub fn to(&mut self, recipient: Recipient<'a, S>) -> &mut Self {
+        self.recipients.push(recipient);
+        self
+    }
+
+    /// Performs the sharing operation with the previously set sharer and recipients.
+    /// It takes the payload, sharer, and recipients, and performs the share operation,
+    /// encrypts the payload and stores it in the sharer's private forest.
+    pub async fn finish(&mut self) -> Result<Rc<PrivateForest>> {
+        if matches!((&self.sharer, self.recipients.len()), (None, 0)) {
+            bail!(ShareError::NoSharerOrRecipients);
+        }
+
+        let sharer = self.sharer.take().unwrap();
+        let recipients = std::mem::take(&mut self.recipients);
+
+        let mut forest = sharer.forest;
+        for recipient in recipients {
+            forest = share::<R>(
+                self.payload,
+                self.count,
+                &sharer.root_did,
+                forest,
+                sharer.store,
+                recipient.exchange_root,
+                recipient.store,
+            )
+            .await?;
+        }
+
+        Ok(forest)
+    }
+}
+
+impl SharePayload {
+    /// Create a share payload from a private fs node.
+    pub async fn from_node(
+        node: &PrivateNode,
+        temporal: bool,
+        forest: Rc<PrivateForest>,
+        store: &mut impl BlockStore,
+        rng: &mut impl RngCore,
+    ) -> Result<(Self, Rc<PrivateForest>)> {
+        let header = node.get_header();
+        let revision_key = header.derive_revision_key();
+        let saturated_name = header.get_saturated_name_with_key(&revision_key);
+        let private_ref =
+            PrivateRef::with_revision_key(Sha3_256::hash(&saturated_name), revision_key.clone());
+
+        let forest = forest
+            .put(saturated_name.clone(), &private_ref, node, store, rng)
+            .await?;
+
+        let payload = if temporal {
+            Self::Temporal(TemporalSharePointer {
+                label: private_ref.saturated_name_hash,
+                revision_key,
+            })
+        } else {
+            Self::Snapshot(SnapshotSharePointer {
+                label: private_ref.saturated_name_hash,
+                content_key: revision_key.derive_content_key(),
+            })
```

Comment:
Couldn't this function re-use functionality from `TemporalSharePointer::from_node` and `SnapshotSharePointer::from_node`?

Reply:
That was the plan. Somehow forgot to do that